File: src/14.longest-common-prefix.cpp
/*
 * @lc app=leetcode id=14 lang=cpp
 *
 * [14] Longest Common Prefix
 */
#include <string>
#include <vector>
using namespace std;
// @lc code=start
class Solution {
 public:
  string longestCommonPrefix(vector<string>& strs) {
    if (strs.size() == 0) {
      return "";
    }
    string prefix = strs.front();

    for (int i = 1; i < strs.size(); i++) {
      prefix = find_prefix(prefix, strs[i]);
    }
    return prefix;
  }

  string find_prefix(string prefix, string s2) {
    int idx = prefix.size();
    while (idx > 0) {
      if (s2.substr(0, idx) == prefix.substr(0, idx)) {
        return prefix.substr(0, idx);
      }
      idx--;
    }
    return "";
  }
};
// @lc code=end

// ["flower","flower","flower","flower"]

File: src/14.longest-common-prefix.js
/*
 * @lc app=leetcode id=14 lang=javascript
 *
 * [14] Longest Common Prefix
 */

// @lc code=start
/**
 * @param {string[]} strs
 * @return {string}
 */
var longestCommonPrefix = function(strs) {
  if (strs.length == 0) {
    return '';
  }

  var prefix = strs[0];

  for (var i = 1; i < strs.length; i++) {
    prefix = getPrefix(prefix, strs[i]);
    if (prefix === '') {
      return prefix;
    }
  }

  return prefix;
};

/**
 * @param {string} strA
 * @param {string} strB
 * @return {string}
 */
var getPrefix = (strA, strB) => {
  var minLength = Math.min(strA.length, strB.length);

  for (var i = minLength; i >= 0; i--) {
    if (strA.substr(0, i) === strB.substr(0, i)) {
      return strA.substr(0, i);
    }
  }
}
// @lc code=end

